### ISSUE TITLE
multi_values fails when parameter does not exist

### ISSUE THREAD

**Issue by tpdownes** [toyokazu/omniauth-shibboleth#19]
I am using gitlab with the omniauth shibboleth plugin. Upon [my request](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/20354), they updated to `omniauth-shibboleth 1.3.0`.

The problem I am trying to solve is that gitlab *really really* wants e-mail to be single-valued and my IdP and AA are both asserting e-mail. GitLab (I believe) uses Devise, so I have tried setting `multi_values` to `first`.

```
  {
    "name"  => "shibboleth",
    "label" => "LIGO-Virgo-KAGRA",
    "args"  => {
      "shib_session_id_field"     => "HTTP_SHIB_SESSION_ID",
      "shib_application_id_field" => "HTTP_SHIB_APPLICATION_ID",
      "uid_field"                 => "HTTP_EPPN",
      "name_field"                => "HTTP_DISPLAYNAME",
      #"info_fields"               => { "email" => lambda {|request_param| request_param.call('HTTP_MAIL').split(/(?<!\\);/).sort[0]} }
      "info_fields"               => { "email" => "HTTP_MAIL" },
      "multi_values"              => "first"
    }
```

This solves the e-mail problem, but it turns out that neither the IdP nor the AA are asserting displayname (`urn:oid:2.16.840.1.113730.3.1.241`). This used to fail gracefully, but it crashes with  `multi_values => first`:

```
NoMethodError (undefined method `gsub' for nil:NilClass):
  lib/gitlab/middleware/multipart.rb:97:in `call'
  lib/gitlab/request_profiler/middleware.rb:14:in `call'
  ee/lib/gitlab/jira/middleware.rb:15:in `call'
  lib/gitlab/middleware/go.rb:17:in `call'
  lib/gitlab/etag_caching/middleware.rb:11:in `call'
  lib/gitlab/middleware/read_only/controller.rb:38:in `call'
  lib/gitlab/middleware/read_only.rb:16:in `call'
  lib/gitlab/request_context.rb:18:in `call'
  lib/gitlab/metrics/requests_rack_middleware.rb:27:in `call'
  lib/gitlab/middleware/release_env.rb:10:in `call'
```

Basically it appears that you are assuming that you have received an object of type string. So, perhaps either a type check needs to be added before calling `gsub` or an exception handler needs to be added?

As it is, a Nil value causes gitlab to return HTTP 500.

You can see from the comment, that I've tried using a lambda call, but I don't think it is working. Your README makes it clear that I shouldn't expect it to work, but I thought I'd try!

**Comment by ghost** [comment]
Could you please provide more information regarding your specific setup? What values do you get if you set `"debug" => "true"`?

I tried to replicate this with GitLab 11.1.0 on Ubuntu 18.04, simulating the Shibboleth headers using `RequestHeader set`, but couldn't get it to work (or rather, fail). The exact config I used was:

```ruby
gitlab_rails['omniauth_providers'] = [{
    "name"  => "shibboleth",
    "label" => "LIGO-Virgo-KAGRA",
    "args"  => {
      "shib_session_id_field"     => "HTTP_SHIB_SESSION_ID",
      "shib_application_id_field" => "HTTP_SHIB_APPLICATION_ID",
      "uid_field"                 => "HTTP_EPPN",
      "name_field"                => "HTTP_DISPLAYNAME",
      "info_fields"               => { "email" => "HTTP_MAIL" },
      "multi_values"              => "first"
    }}]
```

and for Apache (note the absence of a `displayName` entry, to simulate that field not being provided):

```apache
        <Location /users/auth/shibboleth/callback>
                RequestHeader set mail "foo.bar@example.com;fb@example.com"
                RequestHeader set Shib-Session-ID deadbeef
                RequestHeader set Shib-Application-ID default
                RequestHeader set eppn "foo.bar@example.com"
        </Location>
```

On my test VM, I can log in just fine with that config. My display name is set to foo.bar, ie. the local part of the email address.

I am somewhat suspicions that gitlab-shibboleth is the only culprit here. It contains a total of one `.gsub()` that might be problematic, but [that seems to be adequately protected](https://github.com/toyokazu/omniauth-shibboleth/blob/master/lib/omniauth/strategies/shibboleth.rb#L55) from nil values. Also, I cannot see any reference to either Shibboleth or Devise in your backtrace, so while gitlab-shibboleth might be the root cause, there seems to be something else involved as well.

BTW regarding the lambda, yes that indeed cannot work. You can set `multi_values` to a (quoted) lambda expression, but not the field values themselves. That would be logical to add, but I don't see how to distinguish a lambda expression from a weird field name in `OmniAuth::Strategies::Shibboleth.option_handler()` where that would have to be implemented.

**Comment by tpdownes** [comment]
I am on Debian 8 with `gitlab-ee==11.1.1-ee.0`.

I had some suspicion based on traceback that it was outside the `omniauth-shibboleth` gem but the `gsub` made me think your pull request was at fault. Now I believe that the problem is in the automatic creation of a new omniauth user via `gitlab_rails['omniauth_block_auto_created_users'] = false`.

Here's how I tested. Begin by preventing `displayName` assertions from being accepted by creating an `AttributeRule` inside the `AttributeFilterPolicy` for a particular IdP.

```
  <afp:AttributeRule attributeID="displayName">
     <afp:DenyValueRule xsi:type="ANY"/>
  </afp:AttributeRule>
```
Alternatively, one might edit the attribute map to remove a line like this:
```
  <Attribute name="urn:oid:2.16.840.1.113730.3.1.241" id="displayName"/>
```
Either way, one can confirm with `git.example.org/Shibboleth.sso/Session` that there is no `displayName` assertion showing up.

At this point, I agree _authentication for an existing user works just fine_. I now realize that I was actually testing automatic creation of a new user. This is what fails with HTTP 500 and `/var/log/gitlab/gitlab-rails/production.log` shows the traceback in the initial report.

If I then remove `multi_values => first`, you still get HTTP 500 but if you click or refresh, you find that a new user is created with a name set to the name of the account itself. (e.g. `tpdownes`). This may actually be a coincidence that both the name and the account name are being set to the left-hand-side of the e-mail address being asserted.

So I agree that this probably does not reflect a problem with the `omniauth-shibboleth` gem. I will close if you believe so as well.

I also have these related settings enabled though I didn't try turning them on or off. They seem to behave as you would expect for previously-created users. If the assertions exists, the values are properly synced; if the assertion does not exist, no attempt is made to sync.
```
gitlab_rails['omniauth_sync_profile_from_provider'] = [ "shibboleth" ]
gitlab_rails['omniauth_sync_profile_attributes'] = [ "name", "email" ]
```

**Comment by ghost** [comment]
Oops, I meant sign *up* not sign *in*. I deleted the user beforehand, and it was recreated just fine, with display name = username as in your case. (Sign *in* tests very little indeed; it only really needs the `uid_field` to be correct.) I was running it with `gitlab_rails['omniauth_block_auto_created_users'] = false` as well, though I used 11.1.1 CE not EE due to lack of license. I guess I could get an evaluation license if necessary.

Can you reproduce this with two separate sign-ins, restarting the browser and deleting the auto-created user between the tests if necessary? That is, with the `multi_values` setting being the only difference? If so, it is definitely related to omniauth-shibboleth somehow.

**Comment by tpdownes** [comment]
I will do this but I may not be able to do it tomorrow before I have travel next week. So it may be mid-August until you hear back.

I appreciate your effort in creating the PR and @toyokazu for creating a proper release. It allowed me to have GitLab itself use the new gem. Without it, I wouldn't be able to use SAML/Shibboleth for a large set of new users. Sometimes open source works!

**Comment by toyokazu** [comment]
Thank you for your discussions.
About the lambda under device environment, if the most of you need it, I'd like to add a new option, force_eval_option_fields. I just wrote a prototype implementation of :force_eval_option_fields in add_force_eval_option_fields branch.

https://github.com/toyokazu/omniauth-shibboleth/tree/add_force_eval_option_fields

The code is not tested in the real environment. So, if possible, would you try it in your GitLab environment? You can try it by cloning and checking out the branch.

```
git clone https://github.com/toyokazu/omniauth-shibboleth.git
cd omniauth-shibboleth
git checkout add_force_eval_option_fields
rake build
gem install pkg/omniauth-shibboleth-1.3.1.gem
```

Now you can try the latest version 1.3.1 in your environment.
Though this new option may kill the value of the multi_values option, I still keep it for backward compatibility. How to use the option is also added to README.md (using Symbol to distinguish attribute names and lambda descriptions).
If you like it, I'd like to merge it to master and public gem.

Best Regards
